File: android/rule-core/src/main/kotlin/world/zcn/pricereminder/core/MarketScanner.kt
package world.zcn.pricereminder.core

import java.math.BigDecimal
import java.util.ArrayDeque

class MarketScanner {
    private data class Sample(val eventTime: Long, val price: BigDecimal, val priceText: String)
    private data class SymbolState(
        var lastEventTime: Long,
        val rise: ArrayDeque<Sample> = ArrayDeque(),
        val fall: ArrayDeque<Sample> = ArrayDeque(),
    )

    private val states = mutableMapOf<String, MutableMap<String, SymbolState>>()

    @Synchronized
    fun retainRules(ids: Set<String>) { states.keys.retainAll(ids) }

    @Synchronized
    fun resetAll() = states.clear()

    @Synchronized
    fun evaluate(rule: MarketAlertRule, current: PricePoint): List<AlertTrigger> {
        if (!rule.enabled || current.price <= BigDecimal.ZERO) return emptyList()
        val bySymbol = states.getOrPut(rule.id) { mutableMapOf() }
        val sample = Sample(current.eventTime, current.price, current.priceText)
        val state = bySymbol.getOrPut(current.symbol) {
            SymbolState(current.eventTime).apply {
                rise.add(sample)
                fall.add(sample)
            }
        }
        if (current.eventTime < state.lastEventTime) return emptyList()
        val cutoff = current.eventTime - rule.windowMinutes * 60_000L
        state.rise.purge(cutoff)
        while (state.rise.isNotEmpty() && state.rise.last().price >= sample.price) state.rise.removeLast()
        state.rise.addLast(sample)
        state.fall.purge(cutoff)
        while (state.fall.isNotEmpty() && state.fall.last().price <= sample.price) state.fall.removeLast()
        state.fall.addLast(sample)

        val riseBase = state.rise.first()
        val fallBase = state.fall.first()
        val riseChange = current.price.subtract(riseBase.price).divide(riseBase.price, 16, java.math.RoundingMode.HALF_UP)
            .multiply(BigDecimal("100"))
        val fallChange = current.price.subtract(fallBase.price).divide(fallBase.price, 16, java.math.RoundingMode.HALF_UP)
            .multiply(BigDecimal("100"))
        val result = mutableListOf<AlertTrigger>()
        if (riseChange >= rule.threshold) {
            result += trigger(rule, current, riseBase, TriggerDirection.RISE, riseChange)
            state.rise.reset(sample)
        }
        if (fallChange <= rule.threshold.negate()) {
            result += trigger(rule, current, fallBase, TriggerDirection.FALL, fallChange)
            state.fall.reset(sample)
        }
        state.lastEventTime = current.eventTime
        return result
    }

    private fun ArrayDeque<Sample>.purge(cutoff: Long) {
        while (isNotEmpty() && first().eventTime < cutoff) removeFirst()
    }

    private fun ArrayDeque<Sample>.reset(sample: Sample) {
        clear()
        add(sample)
    }

    private fun trigger(
        rule: MarketAlertRule, current: PricePoint, baseline: Sample,
        direction: TriggerDirection, change: BigDecimal,
    ) = AlertTrigger(
        rule.id, current.symbol, AlertRuleKind.MARKET_PERCENTAGE, direction, change,
        rule.thresholdText, rule.windowMinutes, null, current.priceText, baseline.priceText, current.eventTime,
    )
}

File: windows/MarketScanner.cs
namespace PriceReminder.Windows;

internal sealed class MarketScanner
{
    private sealed record Sample(long EventTime, decimal Price, string PriceText);
    private sealed class SymbolState(long eventTime)
    {
        public long LastEventTime { get; set; } = eventTime;
        public LinkedList<Sample> Rise { get; } = [];
        public LinkedList<Sample> Fall { get; } = [];
    }

    private readonly Dictionary<Guid, Dictionary<string, SymbolState>> states = [];
    private readonly object gate = new();

    public void RetainRules(IReadOnlySet<Guid> ids)
    {
        lock (gate)
            foreach (var id in states.Keys.Where(id => !ids.Contains(id)).ToArray()) states.Remove(id);
    }

    public void ResetAll()
    {
        lock (gate) states.Clear();
    }

    public IReadOnlyList<AlertTrigger> Evaluate(MarketAlertRule rule, PricePoint current)
    {
        lock (gate) return EvaluateLocked(rule, current);
    }

    private IReadOnlyList<AlertTrigger> EvaluateLocked(MarketAlertRule rule, PricePoint current)
    {
        if (!rule.Enabled || current.Price <= 0) return [];
        if (!states.TryGetValue(rule.Id, out var bySymbol)) states[rule.Id] = bySymbol = new(StringComparer.OrdinalIgnoreCase);
        var sample = new Sample(current.EventTime, current.Price, current.PriceText);
        if (!bySymbol.TryGetValue(current.Symbol, out var state))
        {
            state = new SymbolState(current.EventTime);
            state.Rise.AddLast(sample);
            state.Fall.AddLast(sample);
            bySymbol[current.Symbol] = state;
        }
        if (current.EventTime < state.LastEventTime) return [];
        var cutoff = current.EventTime - rule.WindowMinutes * 60_000L;
        AppendMinimum(state.Rise, sample, cutoff);
        AppendMaximum(state.Fall, sample, cutoff);
        var riseBase = state.Rise.First!.Value;
        var fallBase = state.Fall.First!.Value;
        var riseChange = (current.Price - riseBase.Price) / riseBase.Price * 100m;
        var fallChange = (current.Price - fallBase.Price) / fallBase.Price * 100m;
        var result = new List<AlertTrigger>();
        if (riseChange >= rule.Threshold)
        {
            result.Add(Trigger(rule, current, riseBase, TriggerDirection.Rise, riseChange));
            Reset(state.Rise, sample);
        }
        if (fallChange <= -rule.Threshold)
        {
            result.Add(Trigger(rule, current, fallBase, TriggerDirection.Fall, fallChange));
            Reset(state.Fall, sample);
        }
        state.LastEventTime = current.EventTime;
        return result;
    }

    private static void AppendMinimum(LinkedList<Sample> values, Sample sample, long cutoff)
    {
        Purge(values, cutoff);
        while (values.Last is { } last && last.Value.Price >= sample.Price) values.RemoveLast();
        values.AddLast(sample);
    }

    private static void AppendMaximum(LinkedList<Sample> values, Sample sample, long cutoff)
    {
        Purge(values, cutoff);
        while (values.Last is { } last && last.Value.Price <= sample.Price) values.RemoveLast();
        values.AddLast(sample);
    }

    private static void Purge(LinkedList<Sample> values, long cutoff)
    {
        while (values.First is { } first && first.Value.EventTime < cutoff) values.RemoveFirst();
    }

    private static void Reset(LinkedList<Sample> values, Sample sample)
    {
        values.Clear();
        values.AddLast(sample);
    }

    private static AlertTrigger Trigger(
        MarketAlertRule rule, PricePoint current, Sample baseline, TriggerDirection direction, decimal change) =>
        new(rule.Id, current.Symbol, AlertRuleKind.MarketPercentage, direction, change,
            rule.ThresholdText, rule.WindowMinutes, null, current.PriceText, baseline.PriceText, current.EventTime);
}
